### ISSUE TITLE
Allow enum declarations to have duplicate members

### ISSUE THREAD

**Issue by rdhananjaya** [ballerina-platform/ballerina-spec#830]
**Description:**

```ballerina
public enum LiftStatus {
    OPEN,
    CLOSED,
    HOLD
}

public enum TrailStatus {
  OPEN,
  CLOSED
}
```

Since both `LiftStatus.OPEN` and `TrailStatus.OPEN` are referring to the same constant string `"OPEN"` should we allow same constant to be declared in multiple enum declarations? I don't think we can do this when the user override the const string `OPEN = "other thing"`.

**Suggested Labels:**

   
**Code sample that shows issue:**

**Related Issues:**

**Comment by jclark** [comment]
You can't declare a constant twice, even with the same value, so I don't think this is allowed at the moment.

**Comment by hasithaa** [comment]
I can write the same with the following,

```ballerina
public const OPEN = "OPEN";
public const CLOSED = "CLOSED";
public const HOLD = "HOLD";

public type LiftStatus OPEN|CLOSED|HOLD;
public type TrailStatus OPEN|CLOSED;
```

 @jclark I think this is very useful in certain cases. Maybe we can relax the rule, saying you can refer to an existing constant in an enum.

**Comment by shafreenAnfar** [comment]
Yeah, I agree with @hasithaa in terms of the usefulness. We actually came across this while implementing the [snowtooth](https://snowtooth.moonhighway.com/) GraphQL API using Ballerina.

**Comment by jclark** [comment]
Ok, we can discuss how to relax this.

**Comment by aneeshafedo** [comment]
Any update on the delivery date of this improvement? We found this issue as an blocker when generating connectors using Ballerina Open API tool where multiple enums are having same members.

**Comment by jclark** [comment]
@aneeshafedo It's easy to workaround this: just use `const` and `type`.

**Comment by SandaruJayawardana** [comment]
Is this a valid scenario?
```ballerina
public enum LiftStatus {
    CLOSED = "0"
}

public enum TrailStatus {
    CLOSED
}
```
I am thinking of allowing an unique value for same constants and generating an error for scenarios like above.

**Comment by jclark** [comment]
@SandaruJayawardana No, you can't do that. `enum` is just some trivial sugar on top of `const` and `type`.

**Comment by rdhananjaya** [comment]
Should we allow the same value to be set to a single `const` with in different `enum` declarations.

```ballerina
public enum LiftStatus {
    CLOSED = "CLOSED"
}

public enum TrailStatus {
    CLOSED = "CLOSED"
}
```

**Comment by jclark** [comment]
@rdhananjaya Sure, that's fine.

**Comment by jclark** [comment]
What do we want to say about metadata on the duplicate members?

**Comment by hasithaa** [comment]
Having different metadata on the duplicate members can be a valid use case.

```ballerina
enum Numbers {
    # Number one
    ONE,
    # Number two
    TWO,
    # Number three
    THREE
}

enum State {
    # State one
    ONE,
    # State two
    TWO
}
```

This is going to be a problem because we have a single const definition for that member. 

I think having different metadata is a compile-time error, is a better solution, given that this case is rare.

**Comment by jclark** [comment]
I am not sure "different" is well-defined for metadata.  We can have readonly values that do not support `==`. Also I don't see the point of making people put the same metadata twice.

I would have thought it would be better to say that only one of duplicates can have metadata.

**Comment by jclark** [comment]
There are a couple of problems with this:

- it leads to the duplicate symbols being associated with multiple source locations in the module, which is not something that happens elsewhere in Ballerina and has wider implications ( e.g. where does VS Code "Goto definition" go to?)
- there is no really satisfactory way to deal with metadata

A better approach would be to extend the *T syntax we already have for record types, object types and classes to enum.

This:

```
enum E { X, Y, *T };
```

would expand into:

```
const X = "X";
const Y = "Y";
type E  X|Y|T;
```

T can be either a string const (effectively allowing for duplicate members) or another enumeration (allowing enumerations to extend other enumerations).

The requirement would be that T is a subtype of string and disjoint from the other members. So in the example above `T&(X|Y)` would have to be empty.

**Comment by jclark** [comment]
Conclusion of discussion:

* @hasithaa will investigate whether we have any internal dependencies on this that cannot be easily removed
* if we have such dependencies, we will put the feature in, with the proviso that only one duplicate can have metadata
* if we do not have such dependencies, we will postpone this feature and consider whether the `*T` approach is a better way to meet the requirement

**Comment by MaryamZi** [comment]
Seems like this may have an impact on the following areas.

- The gRPC protobuf to Ballerina tool currently generates enum declarations for protobuf enumerations. Without support for enum declarations with duplicate members the generated code resulted in errors, but works now. The team mentioned that there was also a feature request specifically for the same (support for proto files with enumerations that have the same members) and this seems to have been used once jBallerina added support for enum declarations with duplicate members. 

- The Ballerina to OpenAPI tool generates enums for Ballerina enum declarations. Right now they have modeled a type definition for a union of string constants different to how they model enums, so simply switching from the enum to a type definition would not preserve the previous behaviour.

Given that the changes required to fix the code/specification generation logic may take some time/effort, I believe it would be better to continue with this feature and introduce the restriction for metadata for now.

**Comment by jclark** [comment]
I decided not to put this in the 2022R1 spec, since it seems clear that the approach we had though would be OK is problematic, and this https://github.com/ballerina-platform/ballerina-spec/issues/830#issuecomment-987516718 is much better.

Since it is too late to remove the feature from jBallerina for Swan Lake, plan is:

- jBallerina has this feature as originally discussed, but it is kept for internal use only and we do not document it
- in next release, we can do the feature right; at that point, we can mention the original feature, but say it is deprecated in favour of the new approach
- we can then convert internal use over to new feature and remove from spec and eventually from implementation
